src/main/scala/io/picnicml/doddlemodel/preprocessing/QuantileDiscretizer.scala

```diff
@@ -0,0 +1,101 @@
+    *
+    *     val featureIndex = FeatureIndex(List(NumericalFeature, CategoricalFeature))
+    *     val x = DenseMatrix(
+    *       List(-1.0, 0.0),
+    *       List(0.0, 1.0),
+    *       List(2.0, 0.0),
+    *       List(5.0, 0.0)
+    *     )
+    *     // equivalently, DenseVector(4) could be used
+    *     val bucketCounts = 4
+    *     val discretizer = QuantileDiscretizer(bucketCounts, featureIndex)
+    *     // Note: no fitting required
+    *     val xTransformed = ev.transform(discretizer, x)
+    *   }}}
+    */
+  def apply(bucketCount: Int, featureIndex: FeatureIndex): QuantileDiscretizer = {
+    val numNumeric = featureIndex.numerical.columnIndices.length
+    val bucketCountExtended: DenseVector[Double] = DenseVector.fill(numNumeric) { bucketCount.toDouble }
+    QuantileDiscretizer(bucketCountExtended, featureIndex)
+  }
+
+  implicit lazy val ev: Transformer[QuantileDiscretizer] = new Transformer[QuantileDiscretizer] {
+
+
+
+    @inline override def isFitted(model: QuantileDiscretizer): Boolean = model.quantiles.isDefined
+
+    override def fit(model: QuantileDiscretizer, x: Features): QuantileDiscretizer = {
+      val xCopy = x.copy
+      val discreteRangeArrays: Seq[Seq[(Double, Double)]] = model.featureIndex.numerical.columnIndices.zipWithIndex.map {
+        case (colIndex, bucketCountsIndex) =>
+          val colArray = xCopy(::, colIndex).toScalaVector.sorted
+          computeQuantiles(colArray, model.bucketCounts, bucketCountsIndex)
+      }
+      model.copy(quantiles = Some(discreteRangeArrays))
+    }
+
+    override protected def transformSafe(model: QuantileDiscretizer, x: Features): Features = {
+      val xCopy = x.copy
+      model.featureIndex.numerical.columnIndices.zipWithIndex.foreach {
+        case (colIndex, bucketsIndex) =>
+          val buckets = model.quantiles.getOrBreak(bucketsIndex)
+          (0 until xCopy.rows).foreach {
+            rowIndex => xCopy(rowIndex, colIndex) = buckets
+              .indexWhere({
+                case (lowerBound, upperBound) => lowerBound <= xCopy(rowIndex, colIndex) && xCopy(rowIndex, colIndex) <= upperBound
+              })
+              .toDouble
+          }
+      }
+      xCopy
+    }
+  }
+
+  private def computeQuantiles(target: Seq[Double], bucketCounts: DenseVector[Double], index: Int): Seq[(Double, Double)] = {
+    val bucketCount = bucketCounts(index).toInt
+    val binPercentileWidth = 1.0 / bucketCount
+    val targetArray = target.toArray
+    // NOTE: Adds binPercentileWidth to make the range inclusive of 1
+    val percentileRangePairs = Range.BigDecimal(0, 1.0+(binPercentileWidth/2), binPercentileWidth).map(_.toDouble).sliding(2).toSeq
+    val rangePairs = percentileRangePairs.map {
+      case Seq(lowerBound, upperBound) =>
+        (DescriptiveStats.percentileInPlace(targetArray, lowerBound), DescriptiveStats.percentileInPlace(targetArray, upperBound))
```

Comment:
What I don't like about this is that it calls `percentileInPlace` twice for most of the `*Bound`, once when the same value is `upperBound` and once when it is `lowerBound`. I would suggest computing percentiles on single values, and then extend the sequence of doubles to the sequence of pairs. WDYT?

Reply:
Fair enough, I believe I can fix this by just moving the `sliding(2)`